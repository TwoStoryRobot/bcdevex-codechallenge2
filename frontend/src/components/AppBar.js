import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Toolbar from '@material-ui/core/Toolbar'
import MAppBar from '@material-ui/core/AppBar'
import Typography from '@material-ui/core/Typography'
import IconButton from '@material-ui/core/IconButton'
import Menu from '@material-ui/core/Menu'
import ListItemAvatar from '@material-ui/core/ListItemAvatar'
import ListItem from '@material-ui/core/ListItem'
import styled from 'styled-components'
import ListItemIcon from '@material-ui/core/ListItemIcon'
import ListItemText from '@material-ui/core/ListItemText'
import Fade from '@material-ui/core/Fade'
import EditIcon from 'mdi-react/EditIcon'
import SignOutIcon from 'mdi-react/LogoutIcon'
import Avatar from 'react-avatar'

const Grow = styled.div`
  flex-grow: 1;
`

const MenuHeader = styled(ListItem)`
  &:focus {
    outline: none;
  }
`

export default class AppBar extends Component {
  state = {
    anchorEl: null
  }

  handleMenuOpen = event => {
    this.setState({ anchorEl: event.currentTarget })
  }

  handleMenuClose = () => {
    this.setState({ anchorEl: null })
  }

  handleMenuItemClick = cb => () => {
    this.handleMenuClose()
    cb()
  }

  render() {
    const { anchorEl } = this.state
    const { title, avatar, name, onEdit, onSignOut } = this.props
    const isMenuOpen = !!anchorEl

    const renderMenu = (
      <Menu
        anchorEl={anchorEl}
        anchorOrigin={{ vertical: 'top', horizontal: 'right' }}
        transformOrigin={{ vertical: 'top', horizontal: 'right' }}
        open={isMenuOpen}
        onClose={this.handleMenuClose}
        data-testid="profile-menu">
        <MenuHeader divider>
          <ListItemAvatar>
            <Avatar round size="40" src={avatar} name={name} />
          </ListItemAvatar>
          <ListItemText primary={name} />
        </MenuHeader>
        <ListItem
          button
          onClick={this.handleMenuItemClick(onEdit)}
          data-testid="edit">
          <ListItemIcon>
            <EditIcon />
          </ListItemIcon>
          <ListItemText inset primary="Edit profile" />
        </ListItem>
        <ListItem
          button
          onClick={this.handleMenuItemClick(onSignOut)}
          data-testid="signout">
          <ListItemIcon>
            <SignOutIcon />
          </ListItemIcon>
          <ListItemText inset primary="Sign out" />
        </ListItem>
      </Menu>
    )

    return (
      <React.Fragment>
        <MAppBar>
          <Toolbar>
            <Typography variant="title" color="inherit" noWrap>
              {title}
            </Typography>
            <Grow />
            {avatar &&
              name && (
                <div>
                  <IconButton
                    onClick={this.handleMenuOpen}
                    data-testid="avatar-button">
                    <Fade in={!isMenuOpen}>
                      <Avatar
                        round
                        size="40"
                        src={avatar}
                        name={name}
                        open={isMenuOpen}
                      />
                    </Fade>
                  </IconButton>
                </div>
              )}
          </Toolbar>
        </MAppBar>
        {renderMenu}
      </React.Fragment>
    )
  }
}

AppBar.propTypes = {
  /** Title to display */
  title: PropTypes.string,
  /** URL of avatar */
  avatar: PropTypes.string,
  /** Name of the user to display in the profile menu */
  name: PropTypes.string,
  /** Handler called when the `Edit` menu button is pressed */
  onEdit: PropTypes.func,
  /** Handler called when the `Sign out` menu button is pressed */
  onSignOut: PropTypes.func
}

AppBar.defaultProps = {
  title: '',
  avatar: '',
  name: ''
}
