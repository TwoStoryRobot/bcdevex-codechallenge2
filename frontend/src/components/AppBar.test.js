import React from 'react'
import { cleanup, render, fireEvent } from 'react-testing-library'

import AppBar from './AppBar'

import 'jest-dom/extend-expect'

afterEach(cleanup)

describe('AppBar', () => {
  it('should match the snapshot', () => {
    const { container } = render(
      <AppBar
        title="Code Challenge"
        avatar="https://api.adorable.io/avatars/285/abott@adorable.png"
        name="Jane Doe"
      />
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should open the menu when the avatar is clicked', () => {
    const { getByTestId, queryByTestId } = render(
      <AppBar
        avatar="https://api.adorable.io/avatars/285/abott@adorable.png"
        name="Jane Doe"
      />
    )

    expect(queryByTestId('profile-menu')).toBeNull()

    fireEvent.click(getByTestId('avatar-button'))

    expect(getByTestId('profile-menu')).toBeVisible()
  })

  it('should close the menu when an action is selected', () => {
    const { getByTestId, queryByTestId } = render(
      <AppBar
        avatar="https://api.adorable.io/avatars/285/abott@adorable.png"
        name="Jane Doe"
      />
    )

    fireEvent.click(getByTestId('avatar-button'))
    fireEvent.click(getByTestId('edit'))

    expect(queryByTestId('profile-menu').firstChild).toHaveStyle('opacity: 0')
  })

  it('should fire the proper menu actions', () => {
    const edit = jest.fn()
    const signout = jest.fn()

    const { getByTestId } = render(
      <AppBar
        avatar="https://api.adorable.io/avatars/285/abott@adorable.png"
        name="Jane Doe"
        onEdit={edit}
        onSignOut={signout}
      />
    )

    // The menu has to be opened for each test because it closes automatically when
    // a MenuItem is selected
    const openMenu = () => fireEvent.click(getByTestId('avatar-button'))

    openMenu()
    fireEvent.click(getByTestId('edit'))
    expect(edit).toHaveBeenCalled()

    openMenu()
    fireEvent.click(getByTestId('signout'))
    expect(signout).toHaveBeenCalled()
  })
})
