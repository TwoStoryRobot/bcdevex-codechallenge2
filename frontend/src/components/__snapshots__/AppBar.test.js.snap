// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBar should match the snapshot 1`] = `
<header
  class="MuiPaper-root-010 MuiPaper-elevation4-016 MuiAppBar-root-01 MuiAppBar-positionFixed-02 MuiAppBar-colorPrimary-08 mui-fixed"
>
  <div
    class="MuiToolbar-root-037 MuiToolbar-regular-039 MuiToolbar-gutters-038"
  >
    <h2
      class="MuiTypography-root-041 MuiTypography-title-047 MuiTypography-colorInherit-060 MuiTypography-noWrap-057"
    />
    <div
      class="sc-bdVaJa kUhAfA"
    />
    <div>
      <button
        tabindex="0"
        class="MuiButtonBase-root-072 MuiIconButton-root-066"
        type="button"
        data-testid="avatar-button"
      >
        <span
          class="MuiIconButton-label-071"
        >
          <div
            class=" sb-avatar"
            style="display: inline-block; vertical-align: middle; width: 40px; height: 40px; border-radius: 100%; font-family: Helvetica, Arial, sans-serif; opacity: 1; -webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
          >
            <div
              class=" sb-avatar__text"
              style="width: 40px; height: 40px; font-size: 13.3333px; line-height: 40px; text-align: center; text-transform: uppercase; color: rgb(255, 255, 255); border-radius: 100%;"
            />
          </div>
        </span>
        <span
          class="MuiTouchRipple-root-075"
        />
      </button>
    </div>
  </div>
</header>
`;
